File: src/main/java/io/protobase/f7/formulas/statistical/MAX.java
package io.protobase.f7.formulas.statistical;

import io.protobase.f7.formulas.AbstractFormula;
import io.protobase.f7.formulas.FormulaName;
import io.protobase.f7.models.GridColumnRowKey;
import io.protobase.f7.utils.Converters;
import io.protobase.f7.utils.Mappers;

import java.util.function.BiFunction;
import java.util.function.Function;
import java.util.stream.Stream;

public class MAX extends AbstractFormula {
  public static FormulaName NAME = FormulaName.MAX;
  public static MAX SELF = new MAX();

  public MAX() {
    super();
  }

  public MAX(Function<Object, Object> lookup, BiFunction<GridColumnRowKey, Object, Object> collateralLookup) {
    super(lookup, collateralLookup);
  }

  @Override
  public Object internal(GridColumnRowKey origin, Object... values) {
    checkAtLeastLength(values.length, 1, NAME);
    return Stream.of(values)
        .map(lookup)
        .flatMap(Mappers::toFlatStreamGridFilterOnlyNumberLiterals)
        .map(Converters::toDouble)
        .reduce(Double::max)
        .orElse(0.0);
  }
}

File: src/main/js/formulas/statistical/MAX.ts
import { isNotNull } from "../../utils/Other";
import { SheetColumnRowKey } from "../../models/common/SheetColumnRowKey";
import { Complex } from "../../models/common/Types";
import { Converters } from "../../utils/Converters";
import { Mappers } from "../../utils/Mappers";
import { Reducers } from "../../utils/Reducers";
import { AbstractFormula } from "../AbstractFormula";
import { FormulaName } from "../FormulaName";

export class MAX extends AbstractFormula {
  static SELF: MAX = new MAX();
  NAME = FormulaName.MAX;

  internal(origin: SheetColumnRowKey, ...values: Array<Complex>) {
    AbstractFormula.checkAtLeastLength(values.length, 1, this.NAME);
    const constructedValues = values
      .map(this.lookup)
      .map(Mappers.flattenGridsToArrays) // 1, 2, [1, "0", "Nope.", null]
      .map(Mappers.filterArrayValuesByIsCoercableToNumeric) // Eg: 1, 2, [1, "0"]
      .map(Mappers.ensureAllAreArrays) // [1], [2], [1, "0"]
      .reduce(Reducers.join) // [1, 2, 1, "0"]
      .filter(isNotNull) // [1, 2, 1, "0"]
      .map(Converters.toNumber); // [1, 2, 1, 0]
    if (constructedValues.length === 0) {
      return 0;
    }
    return constructedValues.reduce(Reducers.max);
  }
}
